Add Azure AD package for remote write

rakshith210 opened the pull request:
This PR contains changes related to Azure AD configuration and Azure AD roundtripper in regard to the following issue:
https://github.com/prometheus/prometheus/issues/11733

Signed-off-by: Rakshith Padmanabha <rapadman@microsoft.com>

LeviHarrison commented on storage/remote/azuread/azuread_config.go:
```diff
@@ -0,0 +1,58 @@
+// for remote write requests to Azure Monitoring Workspace
+type AzureAdConfig struct {
+	// AzureClientId is the clientId of the managed identity that is being used to authenticate.
+	AzureClientId string `yaml:"azure_client_id,omitempty"`
```
nit:
```suggestion
	ClientID string `yaml:"client_id,omitempty"`
```

rakshith210 replied:
done

xy223 replied:
@rakshith210 , will we consider supporting users setting the endpoint url for retrieving aad tokens from the client_id?
We have use cases where instead of using the imds url, we have our own endpoint url for fetching the aad tokens, it'll be really helpful if you could add the endpoint_url, or hostname to the AzureAdConfig.

rakshith210 replied:
Yes I do plan to add AAD based auth in the next iteration. In this PR, we wanted to have changes limited to just one auth (managed identity in this case).

LeviHarrison commented on storage/remote/azuread/azuread_roundtripper.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+// Round tripper adding Azure AD authorization to calls
+type azureAdRoundTripper struct {
```
general nit:
```suggestion
type azureADRoundTripper struct {
```

rakshith210 replied:
done

LeviHarrison commented on storage/remote/azuread/constants.go:
```diff
@@ -0,0 +1,26 @@
+
+package azuread
+
+const (
+	// Clouds
+	AzureChina      = "AzureChina"
+	AzureGovernment = "AzureGovernment"
+	AzurePublic     = "AzurePublic"
+
+	// Audiences
+	INGESTION_CHINA_AUDIENCE      = "https://monitor.azure.cn//.default"
+	INGESTION_GOVERNMENT_AUDIENCE = "https://monitor.azure.us//.default"
+	INGESTION_PUBLIC_AUDIENCE     = "https://monitor.azure.com//.default"
+)
```
This constants can probably just be moved into `token_provider.go`, would make things a bit easier to read.

LeviHarrison commented on storage/remote/azuread/credential.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+// NewTokenCredential return TokenCredential of different kinds like Azure Managed Identity and Azure AD application.
+func NewTokenCredential(cfg *AzureAdConfig) (azcore.TokenCredential, error) {
```
As this package is effectively internal to Prometheus, and the rest of Prometheus would have no reason to call this method directly, I think it should be private.

rakshith210 replied:
Make sense. Made them private

LeviHarrison commented on storage/remote/azuread/credential_mock.go:
```diff
@@ -0,0 +1,38 @@
+	"github.com/stretchr/testify/mock"
+)
+
+// Mock azidentity TokenCredential interface
+type MockCredential struct {
+	mock.Mock
+}
+
+func (m *MockCredential) GetToken(ctx context.Context, options policy.TokenRequestOptions) (azcore.AccessToken, error) {
```
Same thing with the private methods/types here.

rakshith210 replied:
Made the `mockCredential` struct private but `GetToken()` method couldn't be made private  because we are mock the `TokenCredential` interface. https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.3.1/internal/exported#TokenCredential

LeviHarrison commented on storage/remote/azuread/managed_identity_credential.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// Returns new Managed Identity token credential
+func NewManagedIdentityTokenCredential(managedIdentityClientId string) (azcore.TokenCredential, error) {
```
Same thing with the private method here.

rakshith210 replied:
made them private

LeviHarrison commented on storage/remote/azuread/token_provider.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+// TokenProvider is the interface for Credential client
+type TokenProvider interface {
+	GetAccessToken() string
+}
+
+// NewTokenProvider helps to fetch accessToken for different types of credential. This also takes care of
+// refreshing the accessToken before expiry. This accessToken is attached to the Authorization header while making requests.
+func NewTokenProvider(ctx context.Context, cfg *AzureAdConfig, cred azcore.TokenCredential) (TokenProvider, error) {
```
Same thing with the private fields/methods here.

rakshith210 replied:
made them private

LeviHarrison reviewed (commented):
I left a few small comments, but I my one main question so far is can we add OAuth2 as well, and would that be useful? I know for the Azure AD we have both Managed Identity and OAuth.

Also, if you don't mind, could you hook this up to remote write and just make this PR the whole remote write support. I think it would be easier for testing and it shouldn't be much more on top of this (but let me know if it is and then it might make sense to keep it separate).

cstyan commented:
should this live in prometheus/common, the same as the aws sigv4 auth code? https://github.com/prometheus/common/tree/main/sigv4

roidelapluie commented:
The AWS code is in common because it is also used in alertmanager. This code won't be used in alertmanager to the best of my knowledge

rakshith210 commented:
> I left a few small comments, but I my one main question so far is can we add OAuth2 as well, and would that be useful? I know for the Azure AD we have both Managed Identity and OAuth.
> 
> Also, if you don't mind, could you hook this up to remote write and just make this PR the whole remote write support. I think it would be easier for testing and it shouldn't be much more on top of this (but let me know if it is and then it might make sense to keep it separate).

Yes. I do plan to add Azure AD OAuth flow as well. I was thinking of adding it as part of the next PR after this PR is checked in. I wanted to keep this PR concise to just one Auth flow. 

I have changes(fairly minor compared to this change) ready to hook it up to remote write. However, the reason I wanted to take this change too for another PR is because I couldn't directly refer `azuread` package (`github.com/prometheus/prometheus/storage/remote/azuread`) as it is not checked in yet. Let me know if this is fine by you ?

rakshith210 commented on storage/remote/azuread/go.mod:
```diff
@@ -0,0 +1,39 @@
+module github.com/prometheus/prometheus/storage/remote/azuread
```
Removed `go.sum` and `go.mod`

rakshith210 commented on storage/remote/azuread/README.md:
```diff
@@ -0,0 +1,9 @@
+github.com/prometheus/common/azuread module
```
Yes that's correct, my bad. Removed this line.

rakshith210 commented on storage/remote/azuread/Makefile:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2023 The Prometheus Authors
```
done

csmarchbanks commented on storage/remote/azuread/azuread_config.go:
```diff
@@ -0,0 +1,58 @@
+	Cloud string `yaml:"cloud,omitempty"`
+}
+
+// Used to validate config values provided
```
It would be nice if these comments (and in other files) followed the docstring format, e.g. `Validate validates the config values provided.` https://tip.golang.org/doc/comment. Specifically that they should start with the function name and be complete sentences with a full stop.

rakshith210 replied:
Make sense. Updated all the comments to follow the style you suggested.

csmarchbanks commented on storage/remote/azuread/azuread_roundtripper.go:
```diff
@@ -0,0 +1,56 @@
+	return rt, nil
+}
+
+// Sets Auhtorization header for requests
```
Authorization is spelled incorrectly here.

rakshith210 replied:
fixed it

csmarchbanks commented on storage/remote/azuread/token_provider.go:
```diff
@@ -0,0 +1,146 @@
+		case <-tokenProvider.ctx.Done():
+			return nil
+		case <-time.After(tokenProvider.refreshDuration):
+			err := tokenProvider.setAccessToken()
```
Since this done in the background do we need locking inside the setters and getters?

rakshith210 replied:
We are already using `sync/atomic`, which has locking in place and handles the race condition,  to update the values of the accessToken in the `tokenProvider` struct.

```
// setToken handles storing of accessToken value in tokenProvider.
func (tokenProvider *tokenProvider) setToken(token string) {
	var V atomic.Value
	V.Store(token)
	tokenProvider.token = V.Load().(string)
}
```

csmarchbanks commented on config/config.go:
```diff
@@ -826,8 +828,12 @@ func (c *RemoteWriteConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
 
-	if httpClientConfigAuthEnabled && c.SigV4Config != nil {
-		return fmt.Errorf("at most one of basic_auth, authorization, oauth2, & sigv4 must be configured")
+	if httpClientConfigAuthEnabled && (c.SigV4Config != nil || c.AzureADConfig != nil) {
```
Just an idea: I wonder if it is worth counting number of auth options enabled rather than multiple if blocks? That might be easier to understand but I am not sure.

rakshith210 replied:
I thought about this, however, it didn't seem pretty straightforward and felt might get more confusing going forward if some new auth was added.

csmarchbanks commented on storage/remote/azuread/azuread_roundtripper.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Creates round tripper adding Azure AD authorization to calls
+func NewAzureAdRoundTripper(cfg *AzureADConfig, next http.RoundTripper) (http.RoundTripper, error) {
```
This should be `NewAzureADRoundTripper`.

rakshith210 replied:
done

csmarchbanks reviewed (commented):
A few comments, I agree with all of the lint warnings/errors as well and would love to have those cleaned up.

Generally this looks ok to me though!

LeviHarrison commented on storage/remote/azuread/token_provider.go:
```diff
@@ -0,0 +1,146 @@
+		return nil, errors.New("Failed to get access token: " + err.Error())
+	}
+
+	go tokenProvider.periodicallyRefreshClientToken()
```
Instead of having a goroutine do this in the background, could we just check if the token has expired on each request, and if so, update it?

rakshith210 replied:
The idea behind having a goroutine in the background was that we could completely isolate the consumer of the `accessToken` from the refresh logic. Also based to volume of requests we expect in remote write operations, the overhead having background goroutine vs checking for each request seems comparable. Let me know if you would still like me to change the code to check for every request ?

LeviHarrison reviewed (commented):
Looking a lot better!

rakshith210 commented on storage/remote/azuread/azuread_config.go:
```diff
@@ -0,0 +1,58 @@
+// for remote write requests to Azure Monitoring Workspace
+type AzureADConfig struct {
+	// ClientId is the clientId of the managed identity that is being used to authenticate.
+	ClientId string `yaml:"client_id,omitempty"`
```
done

rakshith210 commented on storage/remote/azuread/azuread_config.go:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	if c.ClientId == "" {
+		return fmt.Errorf("must provide a Azure Managed Identity clientId in the Azure AD config")
```
done

rakshith210 commented on storage/remote/azuread/azuread_config.go:
```diff
@@ -0,0 +1,58 @@
+// Used to validate config values provided
+func (c *AzureADConfig) Validate() error {
+	if c.Cloud == "" {
+		return fmt.Errorf("must provide Cloud in the Azure AD config")
```
done

rakshith210 commented on storage/remote/azuread/azuread_config.go:
```diff
@@ -0,0 +1,58 @@
+	_, err := uuid.Parse(c.ClientId)
+
+	if err != nil {
+		return fmt.Errorf("Azure Managed Identity clientId provided is invalid")
```
done

rakshith210 commented on storage/remote/azuread/README.md:
```diff
@@ -0,0 +1,6 @@
+azuread provides a http.RoundTripper that will help attach Azure AD accessToken
+to the requests using the Azure AD library, which is used to retrieve accessToken
+for an Azure Managed Identity.
```
Updated the readme with your suggestion

rakshith210 commented on config/config.go:
```diff
@@ -790,6 +791,7 @@ type RemoteWriteConfig struct {
 	QueueConfig      QueueConfig             `yaml:"queue_config,omitempty"`
 	MetadataConfig   MetadataConfig          `yaml:"metadata_config,omitempty"`
 	SigV4Config      *sigv4.SigV4Config      `yaml:"sigv4,omitempty"`
+	AzureADConfig    *azuread.AzureADConfig  `yaml:"azuread, omitempty"`
```
done

rakshith210 commented on storage/remote/azuread/credential.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+// newTokenCredential return TokenCredential of different kinds like Azure Managed Identity and Azure AD application.
```
done

rakshith210 commented on storage/remote/azuread/credential.go:
```diff
@@ -0,0 +1,36 @@
+func newTokenCredential(cfg *AzureADConfig) (azcore.TokenCredential, error) {
+	var cred azcore.TokenCredential
+	var err error
+	if len(cfg.ClientId) > 0 {
```
removed redundant check

rakshith210 commented on storage/remote/azuread/managed_identity_credential.go:
```diff
@@ -0,0 +1,32 @@
+
+// Returns new Managed Identity token credential
+func newManagedIdentityTokenCredential(managedIdentityClientId string) (azcore.TokenCredential, error) {
+	if len(managedIdentityClientId) > 0 {
```
Same here. Removed it

LeviHarrison commented on storage/remote/azuread/token_provider.go:
```diff
@@ -0,0 +1,145 @@
+type tokenProvider struct {
+	// token is member used to store the current valid accessToken.
+	token string
+	// mu guards access to token
```
nit: all comments should end with a period (there are a few others that should also be fixed)

```suggestion
	// mu guards access to token.
```

rakshith210 replied:
fixed the comments in different places.

LeviHarrison commented on storage/remote/azuread/token_provider.go:
```diff
@@ -0,0 +1,145 @@
+
+// updateRefreshTime handles logic to set refreshTime. The refreshTime is set at half the duration of the actual token expiry.
+func (tokenProvider *tokenProvider) updateRefreshTime(accessToken azcore.AccessToken) error {
+	if len(accessToken.Token) == 0 {
+		return errors.New("Access Token is empty")
+	}
```
I think it would be more logical to have this check take place after the token is fetched in `getToken()`, instead of both here and also in `valid()`

rakshith210 replied:
done

LeviHarrison reviewed (commented):
Looking really good now, only a few more comments (also remember to sign of DCO so tests will run).

One general thing: I know I've asked for smaller file consolidations previously, but I think we want to take all the files in the `azuread` package and combine them into just `azuread.go` and `azuread_test.go`. While separate files are helpful for larger, distinct parts of code, I think in this instance the functionally of the package is both small and related enough that it would be simpler to have everything in one file (well two but you know what I mean).

LeviHarrison commented:
Almost done. I went ahead and tested it out, and here are a few observations:
- Could the `cloud` field just default to `AzurePublic`, with the option of configuring it for different clouds if necessary? I feel like the vast majority of users will be using the public Azure cloud so there's no need to trouble them with an additional field.
- For some reason, the client blocks (I think) when requests are made, stopping any blocks from being sent at all. I don't have time right need to dig through the pprof/debugger to find the root cause, but I will at some point if you haven't already gotten to it.

In addition to those, some things that still need to be worked on in the code:
- Configuration documentation needs to be added in the `configuration.md` file.
- I don't think you responded to my [refresh duration comment](https://github.com/prometheus/prometheus/pull/11944#discussion_r1143946492) or changed anything, so I'm going to assume you're still looking into that.

LeviHarrison commented on docs/configuration/configuration.md:
```diff
@@ -3298,10 +3298,20 @@ sigv4:
   [ <oauth2> ]
 
+# Optional AzureAD configuration.
+# Cannot be used at the same time as basic_auth, authorization, or sigv4.
```
```suggestion
# Cannot be used at the same time as basic_auth, authorization, oauth2, or sigv4.
```

rakshith210 replied:
done

LeviHarrison commented on docs/configuration/configuration.md:
```diff
@@ -3298,10 +3298,20 @@ sigv4:
+# Optional AzureAD configuration.
+# Cannot be used at the same time as basic_auth, authorization, or sigv4.
+azuread:
+  # The Azure Cloud. If blank, the default cloud considered is AzurePublic.
+  [ cloud: <string>]
```
We specify defaults using the following syntax, not in the documentation comment:

```suggestion
  # The Azure Cloud.
  [ cloud: <string> | default = AzurePublic ]
```

rakshith210 replied:
done

LeviHarrison commented on docs/configuration/configuration.md:
```diff
@@ -3298,10 +3298,20 @@ sigv4:
+# Optional AzureAD configuration.
+# Cannot be used at the same time as basic_auth, authorization, or sigv4.
+azuread:
+  # The Azure Cloud. If blank, the default cloud considered is AzurePublic.
```
Could you provide the options for clouds here?

rakshith210 replied:
Added cloud options

LeviHarrison commented on docs/configuration/configuration.md:
```diff
@@ -3298,10 +3298,20 @@ sigv4:
+  # The client id of the Azure managed identity.
+  [ client_id: <string>]
+
+
```
Only one blank line is needed

```suggestion
```

rakshith210 replied:
done

Nexucis commented:
I guess the merge / rebase make a bit mess.

@rakshith210 can you try again. That doesn't look right all these changes are related to your PR :)

rakshith210 commented:
> I guess the merge / rebase make a bit mess.
> 
> @rakshith210 can you try again. That doesn't look right all these changes are related to your PR :)

I just rebased my branch with main. It seems to have fixed it

Nexucis commented:
Perfect thank you !

csmarchbanks reviewed (commented):
Generally the code looks reasonable to me.

I have been trying to test it and running into some issues though, I don't have great Azure experience or access though so might just be doing something wrong.

One thing I notice is that if I have a misconfiguration in Azure auth it looks like Prometheus is not starting at all. Is that expected/desired? I would expect Prometheus to still start up and just writes to fail.

Example of a failed startup for me:
```
ts=2023-05-16T14:44:08.317Z caller=main.go:534 level=info msg="No time or size retention was set so using the default time retention" duration=15d
ts=2023-05-16T14:44:08.318Z caller=main.go:578 level=info msg="Starting Prometheus Server" mode=server version="(version=2.43.0, branch=users/rakshith210/azuread, revision=418e7b399c662af11001c492f2bf447f24cf65aa)"
ts=2023-05-16T14:44:08.319Z caller=main.go:583 level=info build_context="(go=go1.20.4, platform=linux/amd64, user=root@b11ccdb4fa5d, date=20230516-14:02:33, tags=netgo,builtinassets)"
ts=2023-05-16T14:44:08.319Z caller=main.go:584 level=info host_details="(Linux 5.15.0-1037-azure #44~20.04.1-Ubuntu SMP Mon Apr 24 21:52:51 UTC 2023 x86_64 prom (none))"
ts=2023-05-16T14:44:08.320Z caller=main.go:585 level=info fd_limits="(soft=1048576, hard=1048576)"
ts=2023-05-16T14:44:08.320Z caller=main.go:586 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2023-05-16T14:44:08.324Z caller=web.go:562 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
ts=2023-05-16T14:44:08.325Z caller=main.go:1019 level=info msg="Starting TSDB ..."
ts=2023-05-16T14:44:08.328Z caller=head.go:595 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2023-05-16T14:44:08.329Z caller=head.go:676 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=3.3µs
ts=2023-05-16T14:44:08.329Z caller=head.go:684 level=info component=tsdb msg="Replaying WAL, this may take a while"
ts=2023-05-16T14:44:08.332Z caller=tls_config.go:232 level=info component=web msg="Listening on" address=[::]:9090
ts=2023-05-16T14:44:08.332Z caller=tls_config.go:235 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
ts=2023-05-16T14:44:08.334Z caller=head.go:755 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=1
ts=2023-05-16T14:44:08.335Z caller=head.go:755 level=info component=tsdb msg="WAL segment loaded" segment=1 maxSegment=1
ts=2023-05-16T14:44:08.335Z caller=head.go:792 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=40.001µs wal_replay_duration=5.807404ms wbl_replay_duration=100ns total_replay_duration=6.766721ms
ts=2023-05-16T14:44:08.338Z caller=main.go:1040 level=info fs_type=EXT4_SUPER_MAGIC
ts=2023-05-16T14:44:08.338Z caller=main.go:1043 level=info msg="TSDB started"
ts=2023-05-16T14:44:08.339Z caller=main.go:1224 level=info msg="Loading configuration file" filename=prometheus.yml
ts=2023-05-16T14:44:08.348Z caller=main.go:1250 level=error msg="Failed to apply configuration" err="Failed to get access token: ManagedIdentityCredential authentication failed\nGET http://169.254.169.254/metadata/identity/oauth2/token\n--------------------------------------------------------------------------------\nRESPONSE 400 Bad Request\n--------------------------------------------------------------------------------\n{\n  \"error\": \"invalid_request\",\n  \"error_description\": \"Identity not found\"\n}\n--------------------------------------------------------------------------------\nTo troubleshoot, visit https://aka.ms/azsdk/go/identity/troubleshoot#managed-id"
ts=2023-05-16T14:44:08.424Z caller=main.go:878 level=info msg="Stopping scrape discovery manager..."
ts=2023-05-16T14:44:08.424Z caller=main.go:892 level=info msg="Stopping notify discovery manager..."
ts=2023-05-16T14:44:08.424Z caller=manager.go:1009 level=info component="rule manager" msg="Stopping rule manager..."
ts=2023-05-16T14:44:08.424Z caller=manager.go:1019 level=info component="rule manager" msg="Rule manager stopped"
ts=2023-05-16T14:44:08.424Z caller=main.go:929 level=info msg="Stopping scrape manager..."
ts=2023-05-16T14:44:08.424Z caller=main.go:888 level=info msg="Notify discovery manager stopped"
ts=2023-05-16T14:44:08.424Z caller=main.go:874 level=info msg="Scrape discovery manager stopped"
ts=2023-05-16T14:44:08.425Z caller=main.go:921 level=info msg="Scrape manager stopped"
ts=2023-05-16T14:44:08.425Z caller=manager.go:995 level=info component="rule manager" msg="Starting rule manager..."
ts=2023-05-16T14:44:08.430Z caller=notifier.go:606 level=info component=notifier msg="Stopping notification manager..."
ts=2023-05-16T14:44:08.430Z caller=main.go:1150 level=info msg="Notifier manager stopped"
ts=2023-05-16T14:44:08.430Z caller=main.go:1159 level=error err="error loading config from \"prometheus.yml\": one or more errors occurred while applying the new configuration (--config.file=\"prometheus.yml\")"
```

roidelapluie commented:
Prometheus should start indeed in this case . We should not make API calls
while reloading the config

Le mar. 16 mai 2023, 16:51, Chris Marchbanks ***@***.***> a
écrit :

> ***@***.**** commented on this pull request.
>
> Generally the code looks reasonable to me.
>
> I have been trying to test it and running into some issues though, I don't
> have great Azure experience or access though so might just be doing
> something wrong.
>
> One thing I notice is that if I have a misconfiguration in Azure auth it
> looks like Prometheus is not starting at all. Is that expected/desired? I
> would expect Prometheus to still start up and just writes to fail.
>
> Example of a failed startup for me:
>
> ts=2023-05-16T14:44:08.317Z caller=main.go:534 level=info msg="No time or size retention was set so using the default time retention" duration=15d
> ts=2023-05-16T14:44:08.318Z caller=main.go:578 level=info msg="Starting Prometheus Server" mode=server version="(version=2.43.0, branch=users/rakshith210/azuread, revision=418e7b399c662af11001c492f2bf447f24cf65aa)"
> ts=2023-05-16T14:44:08.319Z caller=main.go:583 level=info build_context="(go=go1.20.4, platform=linux/amd64, ***@***.***, date=20230516-14:02:33, tags=netgo,builtinassets)"
> ts=2023-05-16T14:44:08.319Z caller=main.go:584 level=info host_details="(Linux 5.15.0-1037-azure #44~20.04.1-Ubuntu SMP Mon Apr 24 21:52:51 UTC 2023 x86_64 prom (none))"
> ts=2023-05-16T14:44:08.320Z caller=main.go:585 level=info fd_limits="(soft=1048576, hard=1048576)"
> ts=2023-05-16T14:44:08.320Z caller=main.go:586 level=info vm_limits="(soft=unlimited, hard=unlimited)"
> ts=2023-05-16T14:44:08.324Z caller=web.go:562 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
> ts=2023-05-16T14:44:08.325Z caller=main.go:1019 level=info msg="Starting TSDB ..."
> ts=2023-05-16T14:44:08.328Z caller=head.go:595 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
> ts=2023-05-16T14:44:08.329Z caller=head.go:676 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=3.3µs
> ts=2023-05-16T14:44:08.329Z caller=head.go:684 level=info component=tsdb msg="Replaying WAL, this may take a while"
> ts=2023-05-16T14:44:08.332Z caller=tls_config.go:232 level=info component=web msg="Listening on" address=[::]:9090
> ts=2023-05-16T14:44:08.332Z caller=tls_config.go:235 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
> ts=2023-05-16T14:44:08.334Z caller=head.go:755 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=1
> ts=2023-05-16T14:44:08.335Z caller=head.go:755 level=info component=tsdb msg="WAL segment loaded" segment=1 maxSegment=1
> ts=2023-05-16T14:44:08.335Z caller=head.go:792 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=40.001µs wal_replay_duration=5.807404ms wbl_replay_duration=100ns total_replay_duration=6.766721ms
> ts=2023-05-16T14:44:08.338Z caller=main.go:1040 level=info fs_type=EXT4_SUPER_MAGIC
> ts=2023-05-16T14:44:08.338Z caller=main.go:1043 level=info msg="TSDB started"
> ts=2023-05-16T14:44:08.339Z caller=main.go:1224 level=info msg="Loading configuration file" filename=prometheus.yml
> ts=2023-05-16T14:44:08.348Z caller=main.go:1250 level=error msg="Failed to apply configuration" err="Failed to get access token: ManagedIdentityCredential authentication failed\nGET http://169.254.169.254/metadata/identity/oauth2/token\n--------------------------------------------------------------------------------\nRESPONSE 400 Bad Request\n--------------------------------------------------------------------------------\n{\n  \"error\": \"invalid_request\",\n  \"error_description\": \"Identity not found\"\n}\n--------------------------------------------------------------------------------\nTo troubleshoot, visit https://aka.ms/azsdk/go/identity/troubleshoot#managed-id"
> ts=2023-05-16T14:44:08.424Z caller=main.go:878 level=info msg="Stopping scrape discovery manager..."
> ts=2023-05-16T14:44:08.424Z caller=main.go:892 level=info msg="Stopping notify discovery manager..."
> ts=2023-05-16T14:44:08.424Z caller=manager.go:1009 level=info component="rule manager" msg="Stopping rule manager..."
> ts=2023-05-16T14:44:08.424Z caller=manager.go:1019 level=info component="rule manager" msg="Rule manager stopped"
> ts=2023-05-16T14:44:08.424Z caller=main.go:929 level=info msg="Stopping scrape manager..."
> ts=2023-05-16T14:44:08.424Z caller=main.go:888 level=info msg="Notify discovery manager stopped"
> ts=2023-05-16T14:44:08.424Z caller=main.go:874 level=info msg="Scrape discovery manager stopped"
> ts=2023-05-16T14:44:08.425Z caller=main.go:921 level=info msg="Scrape manager stopped"
> ts=2023-05-16T14:44:08.425Z caller=manager.go:995 level=info component="rule manager" msg="Starting rule manager..."
> ts=2023-05-16T14:44:08.430Z caller=notifier.go:606 level=info component=notifier msg="Stopping notification manager..."
> ts=2023-05-16T14:44:08.430Z caller=main.go:1150 level=info msg="Notifier manager stopped"
> ts=2023-05-16T14:44:08.430Z caller=main.go:1159 level=error err="error loading config from \"prometheus.yml\": one or more errors occurred while applying the new configuration (--config.file=\"prometheus.yml\")"
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/11944#pullrequestreview-1428777784>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHHJRIQNGWQXHOEVDVFK3XGOH5ZANCNFSM6AAAAAAUTIRE24>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

rakshith210 commented:
I updated the config structure for azure remote write as follows:
azuread:
  managed_identity:
    client_id: <managed-identity-client-id>

I also fixed the initialization issue faced when wrong config is passed.

csmarchbanks reviewed (commented):
Made some progress testing it, ran into one issue, otherwise I think this looks good!

csmarchbanks commented on docs/configuration/configuration.md:
```diff
@@ -3298,10 +3298,19 @@ sigv4:
 
+# Optional AzureAD configuration.
+# Cannot be used at the same time as basic_auth, authorization, oauth2, or sigv4.
+azuread:
+  # The Azure Cloud. Options are 'AzurePublic', 'AzureChina', or 'AzureGovernment'.
+  [ cloud: <string> | default = AzurePublic ]
+
+  # The client id of the Azure managed identity.
+  [ client_id: <string>]
```
Also, this configuration needs to be updated with the new nested structure.

rakshith210 replied:
done

rakshith210 commented on storage/remote/azuread/azuread.go:
```diff
@@ -0,0 +1,252 @@
+
+// getToken retrieves a new accessToken and stores the newly retrieved token in the tokenProvider.
+func (tokenProvider *tokenProvider) getToken() error {
+	accessToken, err := tokenProvider.credentialClient.GetToken(tokenProvider.ctx, *tokenProvider.options)
```
Updated this! Now passing context from the request

csmarchbanks reviewed (commented):
It looks like there are some lint failures, otherwise :+1: from me!

csmarchbanks reviewed (approved):
:+1: From me, I will wait until next week to merge in case anyone else has comments.

cstyan commented on storage/remote/azuread/README.md:
```diff
@@ -0,0 +1,5 @@
+azuread provides an http.RoundTripper that attaches an Azure AD accessToken
+to remote write requests.
+
+This module is considered internal to Prometheus, without any stability
+guarantees for external usage.
```
nit: should we use some formatting if this is a markdown file?

rakshith210 replied:
Added some basic formatting

cstyan commented on storage/remote/azuread/azuread.go:
```diff
@@ -0,0 +1,247 @@
+)
+
+const (
+	// Clouds.
+	AzureChina      = "AzureChina"
+	AzureGovernment = "AzureGovernment"
+	AzurePublic     = "AzurePublic"
+
+	// Audiences.
+	IngestionChinaAudience      = "https://monitor.azure.cn//.default"
+	IngestionGovernmentAudience = "https://monitor.azure.us//.default"
+	IngestionPublicAudience     = "https://monitor.azure.com//.default"
```
is the list of which of these exist online somewhere? will we have to add more in the future?

rakshith210 replied:
Here is the product catalog where the backend metric store(Managed Prometheus) exists:
https://azure.microsoft.com/en-us/explore/global-infrastructure/products-by-region/?products=monitor

However, these are audiences are per cloud which currently is only the three mentioned above. We will need to add if metrics datastore is made available in new clouds.

cstyan reviewed (commented):
some minor nit comments, I trust Chris to make the decision on merging here :+1:
